### DISCUSSION TITLE
Help with GitHub strategy in regards to RPA development

### DISCUSSION THREAD

**Discussion by mralksndr** [community/community#142977] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

Hello there inspiring minds 🙂

As the title describes I am scouring the internet looking for a GitHub strategy regarding UiPath RPA development.

I am currently working at a place that are investing heavily into the cloud(Azure), and GitHub while sitting with a bunch of RPA devs that want the best setup possible to fit their wishes and needs (sometimes a bit too extreme) 
I've been looking a bit and can't seem to find any posts related to my issue, how to built or configure GitHub strategy for RPA development. What does it look like, what's the best practice, any of you guys out there running it in certain ways or do you have a "one template to rule them all" ?

As of right now it's a mono repo, 3 branches (dev, test, prod(main)) and prod holds all (source of truth) and so this data is pushed to an Orchestrator via a pipeline that packages everything into a nuGet package, and publishes it on the RPA Orchestrator (seems logical in my head, maybe I am wrong?)

I've heard an RPA dev say "why don't we just have a repo for each process that we have" - as of right now there are 40 processes, and it is just growing from here. In my mind it sounds insane having 40+ repos, because of the amount of management, and the constant jumping left and right is just absurd and seems... well, wrong.

Any and all help or insight are much appreciated!
Thank you for your time.

**Comment by thomassiard** [comment]
1. Mono Repo vs. Multi Repo
Keeping everything in a mono repo makes it much easier to manage dependencies and shared components across all processes. As you mentioned, managing 40+ individual repos would be time-consuming and could lead to inconsistent code standards or duplicated efforts. A mono repo allows you to centralize governance and apply consistent CI/CD pipelines, making it easier to manage at scale.
2. Branching Strategy
Your current branch setup (dev, test, prod) aligns well with best practices in RPA and software development. You could also consider using feature branches for major updates or new processes, which can then be merged into the main branches after approval. This approach maintains a controlled flow, allowing you to test and validate individual changes without impacting the primary branches.
3. Modularization Within the Mono Repo
To address concerns about navigating a large mono repo, you could consider structuring the repo by folder for each process (or logical grouping). Each process folder could have its own set of configuration files, assets, and documentation. This way, developers can work on specific processes without sifting through unrelated code.
4. Process-specific Orchestrator Configuration
You might already be doing this, but using GitHub Actions (or another CI/CD tool) to automate the packaging and deployment of processes individually based on changes in their respective folders is a good idea. This would streamline your deployment, ensuring each process is independently versioned and deployed, while still keeping everything centralized.
5. Documentation and Standards
Establishing some standard GitHub guidelines (commit messages, pull request templates, branch naming conventions) can make collaboration smoother, especially as the team grows. Including clear documentation within each process folder will also help with onboarding and code handoffs.

**Reply by mralksndr** [reply]
Thank you so much for you rapid reply Thomas. If you don't mind me asking, do you have much experience in the RPA world ?

**Comment by mralksndr** [comment]
I'd gladly hear from everyone whom have another angle on this. Wanting as much "evidence" to back me up as I got back to the team telling them what's best- and what's not good practice.
